nrf-hal-common/src/usbd/mod.rs

```diff
@@ -0,0 +1,718 @@
+
+            // Check for any finished transmissions.
+            let mut in_complete = 0;
+            let mut out_complete = 0;
+            if regs.events_ep0datadone.read().events_ep0datadone().bit_is_set() {
+                let ep0_state = self.ep0_state.borrow(cs).get();
+                if ep0_state.direction == UsbDirection::In {
+                    // Clear event, since we must only report this once.
+                    regs.events_ep0datadone.reset();
+
+                    in_complete |= 1;
+
+                    // Reset a timeout for the IN transfer
+                    let ep0_state = self.ep0_state.borrow(cs);
+                    let mut state = ep0_state.get();
+                    state.in_transfer_state = TransferState::NoTransfer;
+                    ep0_state.set(state);
+
+                    // Mark the endpoint as not busy
+                    busy_in_endpoints.set(busy_in_endpoints.get() & !1);
+                } else {
+                    // Do not clear OUT events, since we have to continue reporting them until the
+                    // buffer is read.
+
+                    out_complete |= 1;
+                }
+            }
+            let epdatastatus = regs.epdatastatus.read().bits();
+            for i in 1..=7 {
+                if epdatastatus & (1 << i) != 0 {
+                    // EPDATASTATUS.EPIN[i] is set
+
+                    // Clear event, since we must only report this once.
+                    regs.epdatastatus.write(|w| unsafe { w.bits(1 << i) });
+
+                    in_complete |= 1 << i;
+
+                    // Mark the endpoint as not busy
+                    busy_in_endpoints.set(busy_in_endpoints.get() & !(1 << i));
+                }
+                if epdatastatus & (1 << (i + 16)) != 0 {
+                    // EPDATASTATUS.EPOUT[i] is set
+                    // This flag will be cleared in `read()`
+
+                    out_complete |= 1 << i;
+                }
+            }
+
+            // Setup packets are only relevant on the control EP 0.
+            let mut ep_setup = 0;
+            if regs.events_ep0setup.read().events_ep0setup().bit_is_set() {
+                ep_setup = 1;
+
+                // Reset shorts
+                regs.shorts.modify(|_, w| {
+                    w.ep0datadone_ep0status().clear_bit()
+                });
+            }
+
+            // TODO: Check ISO EP
+
+            if out_complete != 0 || in_complete != 0 || ep_setup != 0 {
+                PollResult::Data {
+                    ep_out: out_complete,
+                    ep_in_complete: in_complete,
+                    ep_setup,
+                }
+            } else {
+                PollResult::None
+            }
+        })
+    }
+
+    fn force_reset(&self) -> usb_device::Result<()> {
+        interrupt::free(|cs| {
+            let regs = self.periph.borrow(cs);
+            regs.usbpullup.write(|w| w.connect().disabled());
+            // TODO delay needed?
```

Comment:
I couldn't really find a conclusive answer about the appropriate delay here, but I think there should be a delay of one millisecond here, to make sure that any ongoing transmission (where the host controls the bus and would not notice the device removing the pullup) ends before the pull-up is reattached.